### ISSUE TITLE
[Bug]: ngram speculative decoding default prompt_lookup_min=2 causes tool-call output corruption on Qwen3-class models with structured output (config-only fix: prompt_lookup_min=8)

### ISSUE THREAD

**Issue by Sandermage** [vllm-project/vllm#40875]
### Summary

When using `--speculative-config '{"method":"ngram","prompt_lookup_min":2,...}'` (the default `prompt_lookup_min`) with a Qwen3-class model AND `tool_choice=auto` requests AND a tools array in the system prompt, **tool-call output is corrupted in ~50% of requests** even on a stack with all known related fixes applied (#40738 Phase 1+2, #36138, #40783, #39055).

The corruption manifests as wrong-token cascades like `<tool_call>\n<<tool_code>\nprint(...)`, `parameter=parameter=name`, `<<argname>...<argvalue>` instead of the correct chat-template format `<tool_call>\n<<function=NAME>\n<parameter=KEY>\nVALUE\n</parameter>...`.

**Config-only workaround (no code changes): set `prompt_lookup_min=8`. Achieves 100% clean tool-call rate (n=30 single-query, 96% n=25 multi-query) on the same hardware/model where default `prompt_lookup_min=2` gave ~50%.**

This is a separate bug class from PR #40738 (GDN state corruption, which we backported and confirmed gives +30-40% improvement). The remaining ~50% residual that the four upstream PRs don't close is the rejection-sampling artifact described here.

### Environment

- vLLM: `0.19.2rc1.dev205+g07351e088` (latest nightly tested)
- Model: `Qwen3.6-35B-A3B-FP8` (hybrid linear-attention + MoE + full-attention layers)
- Hardware: 2× NVIDIA RTX A5000 (Ampere SM 8.6), TP=2
- KV cache: `turboquant_k8v4`
- Speculative config: `{"method":"ngram","num_speculative_tokens":3,"prompt_lookup_max":4,"prompt_lookup_min":2}` (defaults)
- Reasoning parser: `qwen3`
- Tool call parser: `qwen3_coder`
- `chat_template_kwargs.enable_thinking`: `false`
- `--enable-chunked-prefill`, `--enable-prefix-caching`

### Reproducer

```python
import json
import requests

PROMPT = {
    "model": "Qwen/Qwen3.6-35B-A3B-FP8",
    "messages": [{"role": "user", "content": "Get weather for Tokyo"}],
    "tools": [{
        "type": "function",
        "function": {
            "name": "get_weather",
            "parameters": {
                "type": "object",
                "properties": {
                    "city": {"type": "string"},
                    "unit": {"type": "string"},
                },
            },
        },
    }],
    "tool_choice": "auto",
    "max_tokens": 150,
    "chat_template_kwargs": {"enable_thinking": False},
}

# Run 30 times, count clean (parsed JSON without `<<` or `parameter=` artifacts)
clean = 0
for _ in range(30):
    r = requests.post("http://localhost:8000/v1/chat/completions", json=PROMPT).json()
    tc = r["choices"][0]["message"]["tool_calls"]
    args = tc[0]["function"]["arguments"] if tc else ""
    if "<<" not in args and "parameter=" not in args and args.startswith("{"):
        clean += 1
print(f"clean: {clean}/30")
```

Expected with default `prompt_lookup_min=2`: ~15-16 clean (50-53%).
Expected with `prompt_lookup_min=8`: 30 clean (100%).

### How we found it (investigation methodology)

This bug was located after backporting four existing upstream PRs failed to fully close the corruption rate. The investigation timeline:

1. **First hypothesis**: TurboQuant KV cache + cudagraph capture/replay (per noonghunna's six-probe ladder at #40807 / #40831). Disproved by Probe 6 (`cudagraph_mode=NONE` works) and Probe 9 (kernel determinism via bit-equality test).

2. **Second pass**: backported #40768 (z1ying, async-scheduler placeholder fix). No measurable improvement — vLLM auto-disables async_scheduling for ngram method (vllm.py:815) so the buggy code path doesn't execute on our config.

3. **Third pass**: backported #39055 (ZenoAFfectionate, Qwen3 reasoning embedded tool_call). +20% improvement (20% → 40%) — confirmed parser-level fixes were on the right track.

4. **Fourth pass**: backported #40738 (tdoublep, GDN+ngram SSM/conv state corruption). +30-40% improvement (43-70% clean). This is the largest single fix and addresses the GDN state corruption originally reported by @bhaktatejas922 in #39273.

5. **Fifth pass**: backported #36138 (sfbemerk, structured-output spec-decode reasoning-end timing) and #40783 (ExtReMLapin, Qwen3 multi-tool first-occurrence + streaming overlap guard). Combined +3-6% additional improvement (53-56% clean). Both are correct fixes for their respective bug classes; just not our dominant residual.

6. **Token-level tracing pass** (where this bug was discovered): ran the reproducer with `return_token_ids: true` + `logprobs: true`, captured 30 responses, decoded raw token IDs via the Qwen3 tokenizer.

#### What the token traces showed

CLEAN response — parsed as `{"city": "Tokyo", "unit": "Celsius"}`:

```
[ 0] 248058 = '<tool_call>'
[ 1]    198 = '\n'
[ 2]     27 = '<'    ← FIRST `<`
[ 3]     27 = '<'    ← SECOND `<` (the qwen3_coder template uses `<<function=>`)
[ 4]   1628 = 'function'   ← CORRECT word
[ 5]  27362 = '=get'
[ 6]  67017 = '_weather'
...
```

BROKEN response — parsed as `<tool_call>\n<<tool_code>\nprint(get_weather(city='Tokyo'))`:

```
[ 0] 248058 = '<tool_call>'
[ 1]    198 = '\n'
[ 2]     27 = '<'
[ 3]     27 = '<'    ← also two `<` (correct per template)
[ 4]  13766 = 'tool'   ← WRONG token here. Should be 'function' (1628)
[ 5]   4000 = '_code'
[ 6]     29 = '>'
[ 7]    198 = '\n'
[ 8]   1302 = 'print'   ← Python code follows because of wrong [4] token
```

**The model picks the wrong token at exactly position [4]** in the broken cases. Position [4] follows the `[27, 27]` `<<` prefix, and the wrong token is one of: `tool` (token 13766, from the `<tool_call>` system-prompt definition), `parameter` (15704), `argname`, `argkey`, etc.

### Root cause analysis

The chat template defines the tool-call format using XML markers that include `<<function=`, `<parameter=`, `</parameter>`, `</function>`. These tokens appear MULTIPLE TIMES in the system prompt's tool definitions section.

When ngram with `prompt_lookup_min=2` is asked "what comes after `[27, 27]` (two `<`)?", the algorithm searches for any 2-token suffix match in the entire context. It finds matches in:

1. The actual `<<function=...>` template definition (correct match — would draft `function`)
2. The `<tool_call>` opening tag (matches `[27, ...]` partially → can draft `tool`)
3. Any literal `<<` markers in user content
4. Multi-tool definitions (each with its own `<<function=NAME>` line)

The KMP-style longest-prefix-suffix matching in [`vllm/v1/spec_decode/ngram_proposer.py`](https://github.com/vllm-project/vllm/blob/main/vllm/v1/spec_decode/ngram_proposer.py) returns ONE of these matches (the longest, but ties are broken by position) and drafts the next k tokens from there.

The rejection sampler then verifies each drafted token against the target model's logits. For the `function` vs `tool` decision at position [4], the target model has `function` at probability ~0.7 and `tool` at probability ~0.2 (both are valid template-context tokens). The rejection sampler accepts a draft token when `target_prob / draft_prob >= random()`. Since ngram doesn't provide draft probabilities, the comparison effectively reduces to "is target_prob above some threshold". For per-token accept rate ≈ 0.8, the structural ceiling for clean rate becomes:

```
clean_rate ≈ (per_token_accept_rate)^num_speculative_tokens
0.8^3 ≈ 0.51 ← matches our 53% empirical measurement with default config
0.8^1 = 0.80 ← matches our 65% with num_spec=1 (statistical variance)
```

#### Why `prompt_lookup_min=8` fixes it

Setting `prompt_lookup_min=8` requires ngram to find an **8-token** suffix match before drafting. For tool-call output, the 8-token suffix `[27, 27, 1628, 27362, 67017, 29, 198, 27]` (`<<function=get_weather>\n<`) is unique enough that it almost never matches spurious system-prompt fragments — it only matches the FIRST tool definition (where the model is currently reproducing the template) OR doesn't match at all.

Result: ngram drafts almost nothing on tool-call requests, producing close to no-spec correctness. Natural-language repetitions (where 8+ token matches do occur, e.g. repeated phrases in long text) still get speculative speedup.

### Empirical proof (clean-rate progression)

All measurements on Qwen3.6-35B-A3B-FP8, 2× A5000, vLLM dev205+g07351e088, with all four upstream backports (#40738 Phase 1+2, #36138, #40783, #39055) applied. n=20-30 reproducer runs per row.

| Config | Clean rate |
|---|---|
| `prompt_lookup_min=2` (default) | 53% (16/30) |
| `prompt_lookup_min=2`, `num_speculative_tokens=1` | 65% (13/20) |
| **`prompt_lookup_min=8`** (single-query reproducer) | **100% (30/30)** |
| **`prompt_lookup_min=8`** (multi-query reproducer, 5 different queries × 5 trials) | **96% (24/25)** |
| no spec-decode (reference) | 100% |

Latency for 300-token responses: ~1.0s with `prompt_lookup_min=8` vs ~0.9s with default. Tradeoff acceptable for tool-call workloads.

### Working production configuration

```bash
--speculative-config '{"method":"ngram","num_speculative_tokens":3,"prompt_lookup_max":10,"prompt_lookup_min":8}'
```

#### Option A — change `prompt_lookup_min` default

Change the default value of `SpeculativeConfig.prompt_lookup_min` from `2` (current) to `4` or higher. Document that low values trade correctness for speedup on workloads where the prompt contains repetitive XML/JSON template fragments (tool definitions, structured output schemas, few-shot examples).

#### Option B — add per-request opt-out

Add a `speculative_config_override` field to chat completion requests so callers can disable spec decoding for individual requests where correctness is paramount. Existing example: tool-call requests could set `enable_speculative=false` while general text requests keep speculative enabled.

#### Option C — context-aware ngram

Modify [`_find_longest_matched_ngram_and_propose_tokens`](https://github.com/vllm-project/vllm/blob/main/vllm/v1/spec_decode/ngram_proposer.py) to skip matches that would land inside a structured-output template fragment (heuristic: prompt section identified by chat template markers). This is more invasive but addresses the root cause directly.

#### Option D — documentation

At minimum, document in [`docs/features/speculative_decoding.md`](https://github.com/vllm-project/vllm/blob/main/docs/features/speculative_decoding.md) that `prompt_lookup_min=2` is unsuitable for tool-calling workloads and recommend `prompt_lookup_min=4-8` for those cases.

### What we still need

1. **Confirmation on other Qwen3 variants** — we tested only Qwen3.6-35B-A3B-FP8. The same template format (`<<function=`) is used by Qwen3-Coder family, so the bug should reproduce there too. Independent confirmation welcomed.

2. **Confirmation on other tool-call parsers** — we tested only `qwen3_coder`. Hermes / Llama tool-call formats may have different vulnerable positions. Worth testing.

3. **Sweet-spot exploration** — we tested `prompt_lookup_min ∈ {2, 8}`. Values 4, 5, 6 may give better speed/correctness tradeoff. We picked 8 as obviously safe; 4-6 might be sufficient with similar speedup.

4. **Reasoning-on test** — we tested with `enable_thinking=false`. With `enable_thinking=true` the model emits `<think>...</think>` blocks where ngram has even MORE template fragments to match against. Likely worse without strict `prompt_lookup_min`.

5. **Default-change PR reception** — open question whether the maintainers prefer Option A (change default), B (per-request opt-out), C (context-aware ngram), or D (documentation only). Happy to write the actual PR for whichever option is preferred.

### References + acknowledgements

This bug class was located only after backporting four existing fixes that addressed adjacent bug classes. Credit to the authors of those PRs whose work narrowed our search:

- @tdoublep + @bhaktatejas922 — [vllm#40738](https://github.com/vllm-project/vllm/pull/40738) / [vllm#39273](https://github.com/vllm-project/vllm/issues/39273) (GDN+ngram SSM/conv state corruption — gave +30-40% improvement)
- @sfbemerk + @cicirori — [vllm#36138](https://github.com/vllm-project/vllm/pull/36138) / [vllm#34650](https://github.com/vllm-project/vllm/issues/34650) (structured-output spec-decode reasoning-end timing)
- @ExtReMLapin — [vllm#40783](https://github.com/vllm-project/vllm/pull/40783) (Qwen3 reasoning parser fragmented tags + multi-tool)
- @ZenoAFfectionate — [vllm#39055](https://github.com/vllm-project/vllm/pull/39055) (Qwen3 reasoning embedded tool_call XML extraction)
- @z1ying + @sweihub — [vllm#40768](https://github.com/vllm-project/vllm/pull/40768) / [vllm#37159](https://github.com/vllm-project/vllm/issues/37159) (async-scheduler placeholder leakage)
- @noonghunna — [vllm#40807](https://github.com/vllm-project/vllm/issues/40807) / [vllm#40831](https://github.com/vllm-project/vllm/issues/40831) (six-probe ladder isolation methodology that made cross-rig data sharing possible)

Full integrated patch tree + investigation reports: [Sandermage/genesis-vllm-patches @ 852b73f](https://github.com/Sandermage/genesis-vllm-patches/tree/main). Specific files relevant to this bug report:

- [`README.md` v7.13 BREAKTHROUGH section](https://github.com/Sandermage/genesis-vllm-patches/blob/main/README.md#v713-breakthrough-2026-04-25-evening--config-only-fix-achieves-100-clean) — documents the formula + working config
- [`CREDITS.md`](https://github.com/Sandermage/genesis-vllm-patches/blob/main/CREDITS.md) — full attribution table

**Comment by Sunt-ing** [comment]
Reproduced on a single **A800 (Ampere SM 8.0)**, current `main`. The corruption is real, but it does not look like a rejection-sampling issue: at temperature=0 ngram greedy writes the target model's argmax at every verified position regardless of the draft, so spec is token-identical to no-spec by construction. What actually breaks it is **FULL cudagraph capturing the TurboQuant spec-verify attention** (`turboquant_3bit_nc` + ngram + `cudagraph_mode=FULL_AND_PIECEWISE`, temp=0 -> degenerate collapse; `enforce_eager` and `PIECEWISE` stay clean). This is the #40807 / #40914 / #44053 / #42551 TurboQuant-cudagraph stack, not the sampler.

Two practical notes:
- On stock `main`, TurboQuant + FULL cudagraph does not even run without #44053 (workspace-lock assertion in `_decode_attention`).
- `prompt_lookup_min=8` is workload-dependent, not a general fix: on repetitive text (where 8-token suffixes still match and ngram keeps drafting) FULL + min=8 still corrupts here. It reaches ~100% clean only on tool-call prompts whose 8-token suffixes are unique enough to suppress drafting.

Full cross-rig isolation matrix + the observation that #40914 reduces but does not fully close it: posted on #40914. Hardware: A800, driver 580.126.09, torch 2.11+cu130.

**Comment by vadiklyutiy** [comment]
Reproduced on **B200** (not just Ampere) with `Qwen/Qwen3.6-35B-A3B-FP8` on current main `0.26.1rc1.dev1424+gdafbef15a`.

Concurrency is not needed — a single sequential request (batch size 1) is enough.

Server:

```
python -m vllm.entrypoints.openai.api_server \
  --model Qwen/Qwen3.6-35B-A3B-FP8 --max-model-len 8192 \
  --tool-call-parser qwen3_coder --reasoning-parser qwen3 --enable-auto-tool-choice \
  --speculative-config '{"method":"ngram","num_speculative_tokens":3,"prompt_lookup_max":4,"prompt_lookup_min":2}'
```

Client (strictly one request at a time):

```python
import requests

PAYLOAD = {
    "model": "Qwen/Qwen3.6-35B-A3B-FP8",
    "messages": [{"role": "user", "content": "Get weather for Tokyo"}],
    "tools": [{"type": "function", "function": {"name": "get_weather", "parameters": {
        "type": "object", "properties": {"city": {"type": "string"}, "unit": {"type": "string"}}}}}],
    "tool_choice": "auto",
    "max_tokens": 1024,
    "chat_template_kwargs": {"enable_thinking": True},
}

bad = 0
for _ in range(30):
    msg = requests.post("http://localhost:8000/v1/chat/completions", json=PAYLOAD).json()["choices"][0]["message"]
    names = [tc["function"]["name"] for tc in (msg.get("tool_calls") or [])]
    if any(n != "get_weather" for n in names):
        bad += 1
        print(names)
print(f"corrupted: {bad}/30")
```

Corrupted function names such as `get_weather(city="Tokyo")`, `get_weather\n<parameter=city` and `get_weather(city="Tokyo")\n</parameter>` appear immediately.

Batch size 1, thinking on, 20-30 requests per row:

| Config | Corrupted |
|---|---|
| ngram `prompt_lookup_min=2` (default) | 22/30 |
| ngram `prompt_lookup_min=1` | 22/30 |
| ngram `num_speculative_tokens=1` | 6/20 |
| ngram, `--no-enable-prefix-caching` | 20/20 |
| ngram, `--enforce-eager` | 18/20 |
| MTP `num_speculative_tokens=3` | 0/20 |
| no spec decode | 0/30 |
| ngram + dense `Qwen3-30B-A3B` | 0/30 |

Four corrections to the analysis in the report.

The corruption is generation-side, not sampler- or parser-side. With `return_token_ids` the raw text is already broken: `<function=get_weather\ncity="Tokyo"\n</parameter>` (missing `>`), plus duplicated spans inside the think block (`"Tokyo"Tokyo"`) and repeated `</think>\n\n<tool_call>\n<function=get_weather` fragments. Dropped and duplicated tokens are a state/bookkeeping fingerprint, not an accepted-bad-draft one — rejection sampling cannot change the output distribution, so the proposed mechanism cannot produce this.

`prompt_lookup_min=8` does not work for the stated reason. At that threshold ngram almost never matches on short tool-call output, so speculation is effectively off. `prompt_lookup_min=1`, where every step drafted exactly 3 tokens (14226 draft tokens / 4746 drafts = 3.0), corrupts exactly as much as the default.

`--enforce-eager` does not help, so this is not the cudagraph capture/replay path.

It is ngram-specific on hybrid GDN models; MTP on the same model is clean and a dense model with the same ngram config is clean. Note that ngram is force-routed to the V1 model runner (`Model Runner V2 does not yet support ngram/ngram_gpu speculative decoding`), so MTP was also re-checked pinned to V1 via `VLLM_USE_V2_MODEL_RUNNER=0` and stayed clean — the runner is not the variable.

AI assistance was used for this investigation.
